### PR TITLE
Change update notification to install by default

### LINKED ISSUE #3054: Always update managed SDK on startup
Rather than notify of an update being available (#3032), use the notification to allow the user to skip the update.

### PR DESCRIPTION
**Pull request by briandealwis**
Fixes #3054 by installing updates by default unless the user indicates they wish to skip.

new-update-notification

The implementation is a bit tricky.  I had hoped to just use the JFace Window/Dialog's support for setting a return code (i.e., `OK` or `CANCEL`).  Unfortunately Mylyn's `AbstractNotificationPopup` implementation has a bug in its _notification-close_ handling (where the user clicks on the `X` in the upper-right to close the notification) in that [it calls `close()` *before* setting the popup's return code](http://git.eclipse.org/c/mylyn/org.eclipse.mylyn.commons.git/tree/org.eclipse.mylyn.commons.ui/src/org/eclipse/mylyn/commons/ui/dialogs/AbstractNotificationPopup.java#n213), unlike the rest of the code that first sets the return code and then closes the popup. 

Further complicating the situation is that the [fade-out code calls `Shell#close()`](http://git.eclipse.org/c/mylyn/org.eclipse.mylyn.commons.git/tree/org.eclipse.mylyn.commons.ui/src/org/eclipse/mylyn/commons/ui/dialogs/AbstractNotificationPopup.java#n509), which triggers a _ShellClosed_ event, eventually calling into [`Window#handleShellCloseEvent()`](http://git.eclipse.org/c/platform/eclipse.platform.ui.git/tree/bundles/org.eclipse.jface/src/org/eclipse/jface/window/Window.java?h=master#n732) .  That calls `setReturnCode(CANCEL)`.  As a result it's hard to tell whether the notification has faded away (= trigger update) vs the user keying _ESC_ or killing the window (= cancel installation).  I discovered that I could use the `Shell`'s _alpha_ value to differentiate between _faded-away_ or _ESC_/_killing-window_, but that's not enough to differentiate on the _notification-close_.

So this PR instead ignores the _returnCode_ and maintains a separate _installation directive_ based on whether the user explicitly selected an item or if the notification faded away.

Note that I left in support for explicitly selecting an _install_ link just in case we want to tweak the text a bit.

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/CloudSdkUpdateNotification.java [hunk 1]
```diff
@@ -44,17 +73,19 @@ public static void showNotification(
...
   private IWorkbench workbench;
   private CloudSdkVersion sdkVersion;
   private Runnable updateRunnable;
 
-  private CloudSdkUpdateNotification(
+  @VisibleForTesting
+  CloudSdkUpdateNotification(
       IWorkbench wb, CloudSdkVersion currentVersion, Runnable triggerUpdate) {
     super(wb.getDisplay());
+    setBlockOnOpen(false);
```
[the remark this answers is not in the document]
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/CloudSdkUpdateNotification.java]: Oops, that was an experiment and should have been removed — I thought `open()` was blocking, but `AbstractPopupNotification` overrides `open()` and is non-blocking and so doesn't make a difference in practice.

**Comment by briandealwis** [comment]
> But the implementation looks like the second case is an implicit "no update"?

I think convincing stories can be made for either behaviour.  I was influenced by the use of ESC / ✖️  in JFace's `Window` to mean `CANCEL`.  For example, ESC / ✖️  in a dialog typically means _dismiss and don't do anything_. But is this a dialog?

I think you're convincing me that the ESC / ✖️ meaning _make this disappear_. And after all the notification is that installation is about to occur, and the user didn't choose the _skip_ option.

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/CloudSdkUpdateNotification.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/CloudSdkUpdateNotification.java]: Don't forget the code comment `// won't wait` above!

**Comment by elharo** [comment]
The more I think about this the more the UI troubles me. Either:

A) This is a non-problematic update, in which case we should do it automatically and not even bother the user
OR
B) It could be problematic in which case we should ask the user to confirm before doing it. (Opt-in)

here we seem to be in a weird in-between case where we'll do it by default but let the user opt-out, but only temporarily and only if they happen to notice the pop-up and click skip quickly enough.

My gut is that we're in case A, and we should just go ahead and update the managed SDK with no skip option and a less prominent notification. If the user wants a specific version they can use the non-managed SDK instead.

**Comment by elharo** [comment]
I can believe that some users don't want to download this. That's case B above. This PR does not adequately address that concern. To handle that we need to give users a clear, opt-In link for updates. Opt-out by happening to notice a transient pop-up that auto-installs if you don't read and click the link fast enough is not sufficient.

Also, I think the metrics on IntelliJ show that users cancel the initial install, not the updates. This PR is only for updates, not installs. Brian, correct me if I'm wrong about that.

**Comment by briandealwis** [comment]
This PR is for updates.  Our installation happens automatically with no notification or confirmation.

(Our auto-installation only happens if the user doesn't already have a Cloud SDK installed.)

**Comment by elharo** [comment]
Is the user guaranteed to see that preference page before we start downloading and updating something?

**Comment by patflynn** [comment]
I see.  Thanks for clarifying.

Ideally we'd be consistent across install/update.

My bias is to be as informative to the user as possible about data use.  We
take our internet connections for granted, but there are a lot of
environments where auto-downloading hundreds of megs would be undesirable.

**Pull request closed without merging**